src/nvim/os/time.c

```diff
@@ -46,26 +47,45 @@ void os_delay(uint64_t milliseconds, bool ignoreinput)
     }
     LOOP_PROCESS_EVENTS_UNTIL(&main_loop, NULL, (int)milliseconds, got_int);
   } else {
-    os_microdelay(milliseconds * 1000);
+    os_microdelay(milliseconds * 1000u, ignoreinput);
   }
 }
 
-/// Sleeps for a certain amount of microseconds
+/// Sleeps for a certain amount of microseconds.
 ///
 /// @param microseconds Number of microseconds to sleep
-void os_microdelay(uint64_t microseconds)
+/// @param ignoreinput If true, ignore all input (including SIGINT/CTRL-C).
+///                    If false, waiting is aborted on any input.
+void os_microdelay(uint64_t microseconds, bool ignoreinput)
 {
-  uint64_t elapsed = 0;
-  uint64_t ns = microseconds * 1000;  // convert to nanoseconds
+  uint64_t elapsed = 0u;
   uint64_t base = uv_hrtime();
 
+  // Convert microseconds to nanoseconds. If uint64_t would overflow, set
+  // nanoseconds to UINT64_MAX.
+  const uint64_t nanoseconds = (microseconds < UINT64_MAX/1000u)
+                               ? microseconds * 1000u
+                               : UINT64_MAX;
+
   uv_mutex_lock(&delay_mutex);
 
-  while (elapsed < ns) {
-    if (uv_cond_timedwait(&delay_cond, &delay_mutex, ns - elapsed)
-        == UV_ETIMEDOUT)
+  while (elapsed < nanoseconds) {
+
+    // If the key input is ignored, we simply wait the full delay. If not, we
+    // check every 10 milliseconds for input and break the waiting loop if input
+    // is available.
+    const uint64_t nanoseconds_delta = (ignoreinput)
+                                       ? nanoseconds - elapsed
+                                       : MIN(nanoseconds - elapsed, 10000000u);
+
+    if ((uv_cond_timedwait(&delay_cond, &delay_mutex, nanoseconds_delta)
+        == UV_ETIMEDOUT) && (ignoreinput || os_char_avail())) {
```

Comment:
With `ignoreinput == false` and `uv_cond_timedwait(...) == UV_ETIMEDOUT || (!ignoreinput && os_char_avail())` we could leave the loop on the first uninterrupted `10` milliseconds. Input would not be checked.

With the unchanged condition:
If `uv_cond_timedwait()` returns early every time, input would never be checked. So we could wait the full time
and never notice new input.